iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/ActiveRegionScanMergeOperator.java

```diff
@@ -0,0 +1,235 @@
+          if (deduplicatedSet.contains(id)) {
+            continue;
+          }
+          deduplicatedSet.add(id);
+        }
+        buildOneRow(i, curTsBlockRowIndex + row, timeColumnBuilder, valueColumnBuilders);
+      }
+      inputIndex[i] += maxRowCanBuild;
+    }
+    return outputCount ? returnResultIfNoMoreData() : tsBlockBuilder.build();
+  }
+
+  private TsBlock returnResultIfNoMoreData() throws Exception {
+    if (isFinished()) {
+      tsBlockBuilder.reset();
+      TimeColumnBuilder timeColumnBuilder = tsBlockBuilder.getTimeColumnBuilder();
+      ColumnBuilder[] valueColumnBuilders = tsBlockBuilder.getValueColumnBuilders();
+      timeColumnBuilder.writeLong(-1);
+      valueColumnBuilders[0].writeLong(count);
+      tsBlockBuilder.declarePosition();
+      count = -1;
+      return tsBlockBuilder.build();
+    }
+    return null;
+  }
+
+  private void buildOneRow(
+      int i,
+      int curTsBlockRowIndex,
+      TimeColumnBuilder timeColumnBuilder,
+      ColumnBuilder[] valueColumnBuilders) {
+    if (outputCount) {
+      count++;
+    } else {
+      timeColumnBuilder.writeLong(inputTsBlocks[i].getTimeColumn().getLong(curTsBlockRowIndex));
+      for (int j = 0; j < valueColumnBuilders.length; j++) {
+        if (inputTsBlocks[i].getValueColumns()[j].isNull(curTsBlockRowIndex)) {
+          valueColumnBuilders[j].appendNull();
+        } else {
+          valueColumnBuilders[j].writeBinary(
+              inputTsBlocks[i].getValueColumns()[j].getBinary(curTsBlockRowIndex));
+        }
+      }
+      tsBlockBuilder.declarePosition();
+    }
+  }
+
+  @Override
+  protected boolean isEmpty(int tsBlockIndex) {
+    return inputTsBlocks[tsBlockIndex] == null
+        || inputTsBlocks[tsBlockIndex].getPositionCount() == inputIndex[tsBlockIndex];
+  }
+
+  @Override
+  public boolean hasNext() throws Exception {
+    if (finished) {
+      return false;
+    }
+    if (retainedTsBlock != null) {
+      return true;
+    }
+    for (int i = 0; i < inputOperatorsCount; i++) {
+      if (!isEmpty(i)) {
+        return true;
+      } else if (!noMoreTsBlocks[i]) {
+        if (!canCallNext[i] || (children.get(i) != null && children.get(i).hasNextWithTimer())) {
+          return true;
+        } else {
+          noMoreTsBlocks[i] = true;
+          inputTsBlocks[i] = null;
+        }
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public boolean isFinished() throws Exception {
+    if (finished) {
+      return true;
+    }
+    if (retainedTsBlock != null) {
+      return false;
+    }
+
+    finished = true;
+    for (int i = 0; i < inputOperatorsCount; i++) {
+      // has more tsBlock output from children[i] or has cached tsBlock in inputTsBlocks[i]
+      if (!noMoreTsBlocks[i] || !isEmpty(i)) {
+        finished = false;
+        break;
+      }
+    }
+    return finished && count == -1;
+  }
+
+  @Override
+  public long calculateMaxPeekMemory() {
+    return 0;
```

Comment:
this should be calculated according to total devices/timeseries memory footprint which can only be got from fe side.